Shrubensembles

sbuschjaeger opened the pull request:
This pull request implements ShrubEnsembles presented in 

Shrub Ensembles for Online Classification
Sebastian Buschjäger, Sibylle Hess, and Katharina Morik
In Proceedings of the Thirty-Sixth AAAI Conference on Artificial Intelligence (AAAI-22), Jan 2022 
[https://aaai.org/papers/06123-shrub-ensembles-for-online-classification/](https://aaai.org/papers/06123-shrub-ensembles-for-online-classification/)

This is a pure-python reimplementation of the original code (https://github.com/sbuschjaeger/se-online) and hence a bit slower than the results mentioned in the paper.(the original code uses C++). Predictive performance is generally comparable to the numbers mentioned in the paper, but some corner cases seem to differ. This might be due to sorting of the data and/or some corner cases that are handled differently.  Overall, however, the implementations should be comparable. I also implemented a regression version (not discussed in the paper, but easily adaptable).  

Let me know what you think :)

tachyonicClock commented:
Thanks for taking the time to submit a pull request. We appreciate it. It looks like the tests aren't quite passing yet. There's no need to worry too much about "conventional commit compliance"; we can handle that when we merge it.

hmgomes commented:
Hi @sbuschjaeger ! Thanks for the contribution, let's talk a bit more about this during our upcoming meeting (will send you an invitation soon)

sbuschjaeger commented:
I updated the documentation and fixed some minor issues with the test cases. Unit tests now all passed on my machine and the docs are building as intended.

sbuschjaeger commented:
I changed the parameters a bit and now it works on my side. However, the documentation seems to also check if the "example usage" code snippets return the exact same string. For example
```
    Example Usage
    >>> from capymoa.datasets import Fried
    >>> from capymoa.regressor import ShrubsRegressor
    >>> from capymoa.evaluation import prequential_evaluation
    >>> stream = Fried()
    >>> schema = stream.get_schema()
    >>> learner = ShrubsRegressor(schema)
    >>> results = prequential_evaluation(stream, learner, max_instances=1000)
    >>> results["cumulative"].rmse()
    4.355469623948763
```
checks if the string "4.355469623948763" matches exactly. Depending on the machine, this string might be longer or shorter and hence the test will fail.

tachyonicClock commented:
Regarding the example usage, something like this should work:
```python
>>> results["cumulative"].rmse()
4.35...
``` 
Or:
```python
>>> print(f"{ results["cumulative"].rmse():.2f}")
4.36
```

tachyonicClock commented on src/capymoa/classifier/_shrubs_ensemble.py:
```diff
@@ -0,0 +1,306 @@
+        self.l_l2_reg = l_l2_reg
+
+        # If the user specifies a random_state, we will also use it. Otherwise we just use 0
+        # TODO This assume that random_state is a number, but it can also be a numpy.randomRandomState. 
```
Can we remove this todo?

sbuschjaeger replied:
Yes, this TODO can be removed. Initially, I used a dictionary containing the parameters for the scikit-learn model, but then refactored the code to directly use a `{DecisionTreeClassifier, DecisionTreeRegressor}` object combined with `clone`. Note however, that I use the `sk_dt.random_state` in the constructor of the `Classifier` object. I am not sure if this is intended.

tachyonicClock commented on src/capymoa/classifier/_shrubs_ensemble.py:
```diff
@@ -0,0 +1,306 @@
+            self.estimator_weights_ = self.estimator_weights_ - step_size*directions - step_size*node_deriv - step_size*self.l_l2_reg*2*self.estimator_weights_
+
+            # The latest tree should only receive updates in the last round of burn-in. Reset its weight here. This is somewhat arbitrary, but lead to better performance in our initial experiments. 
+            # if i < self.burnin_steps:
```
Is the burn-in still being treated correctly

sbuschjaeger replied:
This is a bit more tricky, as both versions are okay. I re-checked my original c++ implementation (which we used for the experiments in the paper) and there we performed burn-in for all trees (i.e. the version that it is now) equally. Hence, this version is closer to the original experiments. I would keep the code but maybe adept the comment accordingly.

tachyonicClock replied:
I decided to remove the commented code (line:255) and associated comment to be more consistent with your C++ implementation. Let me know if I got this the wrong way around and I'll patch it.

tachyonicClock commented:
:eyes: [act](https://github.com/nektos/act) looks cool. Sorry about github not automatically triggering the builds for you hopefully it will now.

I made some changes that are hopefully helpful:

- fixed ``_sgd_regressor.py`` doctest: it still had that precision issue with python 3.9
- I made ``_ShrubEnsembles`` private since a user should not need to create it directly.
- I removed the todo in the predict function.
- Refactored the docstrings to display slightly nicer on the website:

![image](https://github.com/user-attachments/assets/967ccc07-effa-4a77-8672-bcee29aa29fb)

We are getting really close to being able to merge this. I just have some final questions regarding some commented code and todos :D

sbuschjaeger commented:
Perfect, thank you very much :-) 
If you need anything else, let me know.
